### ISSUE TITLE
New invoice payments aren't saved in history

### ISSUE THREAD

**Issue by EZ0010** [ACINQ/phoenix#255]
After latest  update the wallet doesn't seem to save payment invoices and details in the history, balance update but invoices aren't saved.

I did some transactions (send, receive, generate but not pay invoice) balance update after every transaction but I can't see the invoices.
I was able to see the new invoices after I reinstalled the app and had payment backup option enabled, but new generated invoices wont be displayed after that.

iOS 15.4
Phoenix 1.3.2

**Comment by robbiehanson** [comment]
I don't fully understand. Can you help clarify?

> I did some transactions (send, receive, generate but not pay invoice) balance update after every transaction but I can't see the invoices.

You are saying that you don't see the received payments in your payment history? Your balance is correct, but the payments are missing?

**Comment by EZ0010** [comment]
> You are saying that you don't see the received payments in your payment history? Your balance is correct, but the payments are missing?

Exactly, payment goes through and popup is shown but payment details isn't saved in history.
Here is a demo
https://imgur.com/a/RKxH3vH

**Comment by robbiehanson** [comment]
Bug confirmed on iOS 15.4. It might only affect the latest version of iOS. I am investigating...

**Comment by robbiehanson** [comment]
Developer notes:

With iOS 15.4, Apple updated sqlite to version 3.37.0. This particular version of sqlite broke one of our queries. In particular, it broke this one:

```
 SELECT
     combined_payments.type         AS type,
     combined_payments.id           AS id,
     combined_payments.created_at   AS created_at,
     combined_payments.completed_at AS completed_at,
     payments_metadata.modified_at  AS metadata_modified_at
 FROM (
     SELECT
         2            AS type,
         id           AS id,
         created_at   AS created_at,
         completed_at AS completed_at
     FROM outgoing_payments
 UNION ALL
     SELECT
         1                        AS type,
         lower(hex(payment_hash)) AS id,
         created_at               AS created_at,
         received_at              AS completed_at
     FROM incoming_payments
     WHERE incoming_payments.received_at IS NOT NULL
 ORDER BY created_at DESC
--- ^^^^^ Within a UNION ALL statement, the ORDER BY clause is applied to
--- the combined result set, not within the individual result set.
 ) combined_payments
 LEFT OUTER JOIN payments_metadata ON
     payments_metadata.type = combined_payments.type AND
     payments_metadata.id = combined_payments.id
 LIMIT :limit OFFSET :offset;
```

Prior to sqlite v3.37.0, this query would produce a result that was **properly sorted** by `created_at`.

Starting with v3.37.0, this query would produce a very different result:
- first, all the outgoing_payments, sorted by `rowid`
- second, all the incoming_payments, sorted by `rowid`

So the payments actually were being saved into the database. It's just that they were being displayed in the wrong order. (e.g. a new outgoing payment would be displayed at the end of all the other outgoing payments in the list. And a new incoming payment would be displayed at the very bottom of the list.)

The current solution is to move the `ORDER BY` clause outside the `UNION`:

```
 SELECT [...]
 FROM (
     SELECT [...]
     FROM outgoing_payments
 UNION ALL
     SELECT [...]
     FROM incoming_payments
     WHERE incoming_payments.received_at IS NOT NULL
 ) combined_payments
 LEFT OUTER JOIN payments_metadata ON
     payments_metadata.type = combined_payments.type AND
     payments_metadata.id = combined_payments.id
 ORDER BY created_at DESC
 LIMIT :limit OFFSET :offset;
```

In the future we may create a master `payments` table to increase performance.

**Comment by robbiehanson** [comment]
This should be fixed in v1.3.3, which is now available in the App Store.

**Comment by EZ0010** [comment]
just updated it and it is fixed, thanks a lot 👍
